fdk.go

```diff
@@ -3,94 +3,142 @@ package fdk
+	CallID() string
+
+	// TODO should we unwind all known env var fields as methods here, as well?
+}
+
+// HTTPContext contains all configuration for a function invocation sourced
+// from an http gateway trigger, which will make the function appear to receive
+// from the client request they were sourced from, with no additional headers.
+type HTTPContext interface {
+	Context
+
+	// RequestURL is the request url from the gateway client http request
+	RequestURL() string
+
+	// RequestMethod is the request method from the gateway client http request
+	RequestMethod() string
+}
+
+type baseCtx struct {
+	header http.Header
+	config map[string]string
+	callID string
 }
 
-// Ctx provides access to Config and Headers from fn.
-type Ctx struct {
-	// Header are the unmodified headers as sent to the container, see
-	// HTTPHeader for specific trigger headers
-	Header http.Header
-	// HTTPHeader are the request headers as they appear on the original HTTP request,
-	// for an http trigger.
-	HTTPHeader http.Header
-	Config     map[string]string
-	RequestURL string
-	Method     string
+type httpCtx struct {
+	// XXX(reed): if we embed we won't preserve the original headers. since we have an
+	// interface handy now we could change this under the covers when/if we want... idk
+	baseCtx
+	requestURL    string
+	requestMethod string
+}
+
+func (c baseCtx) Config() map[string]string { return c.config }
+func (c baseCtx) Header() http.Header       { return c.header }
+func (c baseCtx) CallID() string            { return c.callID }
+func (c baseCtx) ContentType() string       { return c.header.Get("Content-Type") }
+
+func (c httpCtx) RequestURL() string    { return c.requestURL }
+func (c httpCtx) RequestMethod() string { return c.requestMethod }
+
+func ctxWithDeadline(ctx context.Context, fnDeadline string) (context.Context, context.CancelFunc) {
+	t, err := time.Parse(time.RFC3339, fnDeadline)
+	if err == nil {
+		return context.WithDeadline(ctx, t)
+	}
+	return context.WithCancel(ctx)
 }
 
-// AddHeader will add a header on the function response, for hot function
-// formats.
+// AddHeader will add a header onto the function response
 func AddHeader(out io.Writer, key, value string) {
-	if resp, ok := out.(*utils.Response); ok {
-		resp.Header.Add(key, value)
+	if resp, ok := out.(http.ResponseWriter); ok {
+		resp.Header().Add(key, value)
 	}
 }
 
-// SetHeader will set a header on the function response, for hot function
-// formats.
+// SetHeader will set a header on the function response
 func SetHeader(out io.Writer, key, value string) {
-	if resp, ok := out.(*utils.Response); ok {
-		resp.Header.Set(key, value)
+	if resp, ok := out.(http.ResponseWriter); ok {
+		resp.Header().Set(key, value)
 	}
 }
 
-// WriteStatus will set the status code to return in the function response, for
-// hot function formats.
+// WriteStatus will set the status code to return in the function response
 func WriteStatus(out io.Writer, status int) {
```

Comment:
Not sure if the code should be able to override the container status (but not sure) - this is different from `SetHeader('Fn-Http-Status', "202")` e.g.

Reply:
it will end up getting unwound to that under `Fn-Intent=httprequest` - for invoke, the right behavior wasn't clear to me (I should have read the other fdks in more detail, perhaps, my b). that would seem to imply the http contract only applies to wrapping up `Fn-Http-H-` but `Fn-Http-Status` should be a thing regardless of the type of trigger/invoke from which it was sourced (and the trigger can do whatever it wants with that, and invoke can set it as the response status code or just leave the header as it was, I think we're doing the latter atm).

Reply:
Can we also change the signature of this method to:
```go
SetHeader(io.Writer, key string, value []string)
```
because Go header is not a mapping of a `string to string`, but `the string to a slice of strings`.

Reply:
semantically this works the same as `http.Header.Set` - there's also `AddHeader(io.Writer, key, value string)` to get at what you're asking - so gonna say no, open to ideas about making headers more accessible I still find this interaction slightly awkward, however we have `http.Handler` wrapper now so I think users that want to set headers should probably use that anyway...